src/cfnlint/rules/resources/codepipeline/CodepipelineStageActions.py

```diff
@@ -0,0 +1,228 @@
+        """Check that artifact counts are within valid ranges."""
+        matches = []
+
+        action_type_id = action.get('ActionTypeId', {})
+        owner = action_type_id.get('Owner')
+        if owner not in self.CONSTRAINTS.keys():
+            return matches
+
+        category = action_type_id.get('Category')
+        if not category:
+            return matches
+
+        provider = action_type_id.get('Provider')
+        if not provider:
+            return matches
+
+        constraints = self.CONSTRAINTS.get(owner, {}).get(category, {}).get(provider, {})
+        if not constraints:
+            return matches
+        artifact_count = len(action.get(artifact_type, []))
+
+        constraint_key = self.KEY_MAP[artifact_type]
+        if isinstance(constraints[constraint_key], tuple):
+            min_, max_ = constraints[constraint_key]
+            if not (min_ <= artifact_count <= max_):
+                message = (
+                    'Action "{action}" declares {number} {artifact_type} which is not in '
+                    'expected range [{a}, {b}].'
+                ).format(
+                    action=action['Name'],
+                    number=artifact_count,
+                    artifact_type=artifact_type,
+                    a=min_,
+                    b=max_
+                )
+                matches.append(RuleMatch(
+                    path + [artifact_type],
+                    message
+                ))
+        else:
+            if artifact_count != constraints[constraint_key]:
+                message = (
+                    'Action "{action}" declares {number} {artifact_type} which is less than '
+                    'expected [{a}].'
+                ).format(
+                    action=action['Name'],
+                    number=artifact_count,
+                    artifact_type=artifact_type,
+                    a=constraints[constraint_key]
+                )
+                matches.append(RuleMatch(
+                    path + [artifact_type],
+                    message
+                ))
+
+        return matches
+
+    def check_owner(self, action, path):
+        """Check that action type owner is valid."""
+        matches = []
+
+        owner = action.get('ActionTypeId', {}).get('Owner')
+        if owner not in self.VALID_OWNER_STRINGS:
+            message = (
+                'For all currently supported action types, the only valid owner '
+                'strings are {owners}'
+            ).format(
+                owners=', '.join(list(self.VALID_OWNER_STRINGS))
+            )
+            matches.append(RuleMatch(
+                path + ['ActionTypeId', 'Owner'],
+                message
+            ))
+
+        return matches
+
+    def check_version(self, action, path):
+        """Check that action type version is valid."""
+        matches = []
+
+        if action.get('ActionTypeId', {}).get('Version') != 1:
```

Comment:
We may want to check that ActionTypeId is a dict and not a list (or some other object) before doing the get on the result.

Going to my other comment if we had tests to validate ActionTypeId is an object we would just ignore a failure here.  Then here we are just making sure its an object before we do a .get (if its a list it will crash).   I always debate generic try/catches here for common errors like this.  At least when your checks aren't validating that specifically you can ignore them.  

But I'm thinking a generic CodePipeline syntax checker would be nice.  Just covering the basic structure and supported types for the elements.

Reply:
Never mind that is in the CloudFormation spec from what I can tell.  So we should be putting in a more generic try/catch on AttributeError for 'AttributeError: 'list_node' object has no attribute 'get'' that you could skip but that would result in cfn-lint not crashing.

Or we would want to test the items are dicts before doing a get.